### ISSUE TITLE
Table header-finding algorithm requires clarification 

### ISSUE THREAD

**Issue by DuffJohnson** [pdf-association/pdf-issues#200]
**Describe the bug**
14.8.4.8.3; the algorithm below Table 371, is less than completely clear.

- In the 2nd paragraph following the table, replace the "/" between "horizontal/vertical" with "and". 
- In the last sentence of the 2nd paragraph following the table, add "In both directions," at the beginning of the sentence.
- In the 3rd bullet of the list following the 2nd para, replace the "/" between "row/column" with "and/or"
- In the 3rd para following the table, strike "and the (implicit or explicit) Scope attribute of the header cell is either Both or Row/Column,"
- In the 3rd para following the table, replace the "/" in "row/column headers" with "or".

**Comment by FrankMittelbach** [comment]
My take:

 -  In the 2nd paragraph following the table, replace "appropriate horizontal/vertical direction"  with "appropriate horizontal and vertical directions"  (guess that's a plural)
 - ok as suggested
 - In the 3rd bullet of the list following the 2nd para, replace "the row or column list that is currently being built".  (Rationale: the algorithm builds list for both column and row in the two searches it makes)
 - I get second thoughts on this one, see below
 - ok

My take on
> When a header cell is found in the search and the (implicit or explicit) Scope attribute of the header cell is either Both or Row/Column, the header cell is appended to the end of the list of row/column headers, resulting in a list of headers ordered from most specific to most general

that the text should not be dropped but rather extended.

I think what is meant really is this (not in the best English):

When a header cell is found in the search and the (implicit or explicit) Scope attribute of the header cell matches the current search direction (i.e., is either Both, or is Row or Column and we are searching for row or column headers), then the header cell is appended to the end of the list of row or column headers, resulting in a list of headers ordered from most specific to most general.

In other words, If we look up the column but find a header cell (without header attribute, because then we stop) and the Scope (wither explicit or implicit) of this header cell is Row and not Column or Both, *then* this is not a header cell we should care about. We should only care about it when it is Both or Column

**Comment by DuffJohnson** [comment]
Concur; thanks, Frank. My suggestion for slightly tweaked text:

When a header cell is found in the search and the (implicit or explicit) **Scope** attribute of the header cell matches the current search direction (i.e., is either Both, or is Row or Column when searching for row or column headers, respectively), then the header cell is appended to the end of the list of row or column headers, resulting in a list of headers ordered from most specific to most general.

**Comment by car222222** [comment]
The preamble paragraph to the algorithm is somewhat strange, ﻿and the first sentence seems to be unnecessary.  It could read more straightforwardly, as follows --

The headers of a cell can be defined explicitly by using the Headers attribute of the cell; if there is no Headers attribute then the headers of a cell shall be deduced by applying the following algorithm:

**Comment by car222222** [comment]
The termination actions in the second two listed items of the algorithm probably also need to be qualified in case the given header cell has a Scope attribute value incompatible with the current stage of the search.  Thus:
Cells with an incompatible value for Scope should not terminate this section of the search (and they should not be added to the headers list being created).
(Note that needs confirmation, but it seems to follow logically.)

In other words, for example, If we are looking up a column but find a header cell  with a Scope attribute value of Row, then this cell should be completely ignored and so not cause this column search to terminate.

**Comment by car222222** [comment]
A question: when adding to a list from the lists in a found cell’s Header attribute, should only that part of the list for the currently being searched direction (row or column) be added to the list being of headers created?   

The current text suggests that both the row and the column list should be added but is this sensible/correct?
Maybe, from the value of the Headers attribute, only the list of cells with the “currently being searched orientation” should be added (rather than the whole list).

**Comment by FrankMittelbach** [comment]
@car222222 are you refering to the original writing or to the suggested rewrites? Because I thought the last 2 points you made are precisely the ones that the rewrites are supposed to address. (it is all a bit difficult to deciper in the issue if only suggested changes are listed but not the full new text).

**Comment by car222222** [comment]
@FrankMittelbach I do not think that there is any direct relationship between my second suggestions/questions and the text concerning the reworded details of dealing with the scope attribute.

After there is a decision about the exact details of the algorithm then we can work on the wording.  My first comment contains a complete wording and the last is at this stage only a question.  I could nevertheless provide some suggested wordings for the various possibilities.

**Comment by DuffJohnson** [comment]
I've attempted to synthesize the comments thus far into the following update to the existing algorithm below Table 371...
---
The headers of a cell can be defined explicitly by using the **Headers** attribute of the cell; if there is no **Headers** attribute then the headers of a cell shall be deduced by applying the following algorithm:

To find headers for any data or header cell, begin from the current cell position and use the current value of **WritingMode** to search towards the first cell in the appropriate horizontal and vertical directions. The search terminates when any of these conditions is reached: 

• the edge of the table is reached 
• a data cell is found after a header cell 
• a header cell has the **Headers** attribute set — the headers that are specified are appended to the row or column list that is current being built

Cells with a value for **Scope** that conflicts with the search's direction should not terminate this section of the search and should therefore not be added to the headers list being created.

EXAMPLE: If searching up a given column reveals a cell with a **Scope** value of _Row_ this cell should not cause this column search to terminate.

When a header cell is found in the search and the (implicit or explicit) **Scope** attribute of the header cell matches the current search direction (i.e., is either _Both_, or is _Row_ or _Column_ when searching for row or column headers, respectively), then the header cell is appended to the end of the list of row or column headers, resulting in a list of headers ordered from most specific to most general.

**Comment by car222222** [comment]
Looks good, Duff.  Thanks.
﻿
To be crystal clear, something like the following could be used:  

  . . . and such a cell (or anything from its Headers attribute) should not be added to the headers list being created.

to replace this:

 . . . and should therefore not be added to the headers list being created.

**Comment by DuffJohnson** [comment]
OK, now updated...
---
The headers of a cell can be defined explicitly by using the Headers attribute of the cell; if there is no Headers attribute then the headers of a cell shall be deduced by applying the following algorithm:

To find headers for any data or header cell, begin from the current cell position and use the current value of WritingMode to search towards the first cell in the appropriate horizontal and vertical directions. The search terminates when any of these conditions is reached:

• the edge of the table is reached
• a data cell is found after a header cell
• a header cell has the Headers attribute set — the headers that are specified are appended to the row or column list that is current being built

Cells with a value for Scope that conflicts with the search's direction should not terminate this section of the search and such a cell (irrespective of its Headers attribute) should not be added to the headers list being created.

EXAMPLE: If searching up a given column reveals a cell with a Scope value of Row this cell should not cause this column search to terminate.

When a header cell is found in the search and the (implicit or explicit) Scope attribute of the header cell matches the current search direction (i.e., is either Both, or is Row or Column when searching for row or column headers, respectively), then the header cell is appended to the end of the list of row or column headers, resulting in a list of headers ordered from most specific to most general.

**Comment by DuffJohnson** [comment]
The PDF/UA TWG believes the 2022-09-01 proposed text, above, is accurate and suggests that the PDF TWG proceed with adoption into forthcoming errata.

**Comment by mrbhardy** [comment]
I would suggest changing:

> The headers of a cell can be defined explicitly by using the Headers attribute of the cell; if there is no Headers attribute then the headers of a cell shall be deduced by applying the following algorithm:

to:

> When a cell does not explicitly identify its headers through a Headers attribute, the following algorithm may be used to determine the cells headers, if any:

Then start the following line:

> Given a data or header cell...

**Comment by DuffJohnson** [comment]
I agree with Matthew's comment modifying the solution proposed last September.

**Comment by petervwyatt** [comment]
Final review by PDF/UA TWG including Matthew's tweaks.

**Comment by DuffJohnson** [comment]
The final text for review/approval in the next PDF/UA TWG Meeting (Sept 19), incorporating @mrbhardy's suggestion:
----
When a cell does not explicitly identify its headers through a Headers attribute, the following algorithm may be used to determine the cell’s headers, if any:

Given a data or header cell, begin from the current cell position and use the current value of WritingMode to search towards the first cell in the appropriate horizontal and vertical directions. The search terminates when any of these conditions is reached:

• the edge of the table is reached
• a data cell is found after a header cell
• a header cell has the Headers attribute set — the headers that are specified are appended to the row or column list that is current being built

Cells with a value for Scope that conflicts with the search's direction should not terminate this section of the search and such a cell (irrespective of its Headers attribute) should not be added to the headers list being created.

EXAMPLE: If searching up a given column reveals a cell with a Scope value of Row this cell should not cause this column search to terminate.

When a header cell is found in the search and the (implicit or explicit) Scope attribute of the header cell matches the current search direction (i.e., is either Both, or is Row or Column when searching for row or column headers, respectively), then the header cell is appended to the end of the list of row or column headers, resulting in a list of headers ordered from most specific to most general.

**Comment by car222222** [comment]
Two small ideas to, I hope, improve the clarity:

1.  add text to this:
"a header cell has the Headers attribute set"
to get:
   a header cell is found that has ...

2.  To make a better link, change the start of the following paragraph from 
   "Cells . . ." 
to:
  However, cells . . .

**Comment by DuffJohnson** [comment]
The PDF/UA TWG incorporated @car222222's suggestions and approved the following text.
-----
When a cell does not explicitly identify its headers through a Headers attribute, the following algorithm may be used to determine the cell’s headers, if any:

Given a data or header cell, begin from the current cell position and use the current value of WritingMode to search towards the first cell in the appropriate horizontal and vertical directions. The search terminates when any of these conditions is reached:

• the edge of the table is reached
• a data cell is found after a header cell
• a header cell is found that has the Headers attribute set — the headers that are specified are appended to the row or column list that is current being built

However, cells with a value for Scope that conflicts with the search's direction should not terminate this section of the search and such a cell (irrespective of its Headers attribute) should not be added to the headers list being created.

EXAMPLE: If searching up a given column reveals a cell with a Scope value of Row this cell should not cause this column search to terminate.

When a header cell is found in the search and the (implicit or explicit) Scope attribute of the header cell matches the current search direction (i.e., is either Both, or is Row or Column when searching for row or column headers, respectively), then the header cell is appended to the end of the list of row or column headers, resulting in a list of headers ordered from most specific to most general.

**Comment by mrbhardy** [comment]
I like the new text @DuffJohnson
